Implement optional ring buffer for container logs

cpuguy83 opened the pull request:
This allows the user to set a logging mode to "blocking" (default), or
"non-blocking", which uses the ring buffer as a proxy to the real log
driver.

This allows a container to never be blocked on stdio at the cost of
dropping log messages.

Introduces a new flag to `docker run` and `dockerd`, `--log-mode`, which
takes a value of "blocking", or "non-blocking". I chose not to implement
this as a bool since it is difficult to determine if the mode was set to
false vs just not set... especially difficult when merging the default
daemon config with the container config.

One thing to think about here is `logger.Message` could be any size, it just depends on how long the log line was... so it could be almost nothing, or it could be several megabytes which could cause problems in just blindly buffering N number of messages in the ring buffer.
We could keep track of this, but this could be tricky and introduces locking into the implementation.

cpuguy83 committed:
Implement optional ring buffer for container logs

This allows the user to set a logging mode to "blocking" (default), or
"non-blocking", which uses the ring buffer as a proxy to the real log
driver.

This allows a container to never be blocked on stdio at the cost of
dropping log messages.

Introduces 2 new log-opts that works for all drivers, `log-mode` and
`log-size`. `log-mode` takes a  value of "blocking", or "non-blocking"
I chose not to implement this as a bool since it is difficult to
determine if the mode was set to false vs just not set... especially
difficult when merging the default daemon config with the container config.
`log-size` takes a size string, e.g. `2MB`, which sets the max size
of the ring buffer. When the max size is reached, it will start
dropping log messages.

```
BenchmarkRingLoggerThroughputNoReceiver-8           	2000000000	        36.2 ns/op	 856.35 MB/s	       0 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputWithReceiverDelay0-8   	300000000	       156 ns/op	 198.48 MB/s	      32 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay1-8        	2000000000	        36.1 ns/op	 857.80 MB/s	       0 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay10-8       	1000000000	        36.2 ns/op	 856.53 MB/s	       0 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay50-8       	2000000000	        34.7 ns/op	 894.65 MB/s	       0 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay100-8      	2000000000	        35.1 ns/op	 883.91 MB/s	       0 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay300-8      	1000000000	        35.9 ns/op	 863.90 MB/s	       0 B/op	       0 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay500-8      	2000000000	        35.8 ns/op	 866.88 MB/s	       0 B/op	       0 allocs/op
```

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

justincormack commented:
Seems like one reasonable solution. I wonder if the ring buffer should be a fixed size in bytes though (which can be specified). Then this would carry over simply to a compatible interfaces via a pipe, where you could set the pipe buffer size and discard if no space. There are some issues about message boundaries though, you could use datagrams and allow truncation.

LK4D4 commented:
@cpuguy83 maybe embed it to log-opt?

cpuguy83 commented:
> @cpuguy83 maybe embed it to log-opt?

@LK4D4 The reason I didn't want to do that is `log-opt` tends to be options for the driver, where as this is changing how docker itself handles logs.

cpuguy83 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,122 @@
+	default:
+		select {
+		case <-r.ctx.Done():
+		case <-r.queue: // Make room on the queue
```
May need to add a lock to protect against another log message coming in right after this but before we get a chance to queue below.

LK4D4 replied:
I'd better try to implement messages ring buffer as a separate data structure with tests and stuff. Probably that was your plan after design review.

LK4D4 commented:
I agree with @justincormack about specifying size. However, computing size might be challenging. The number of messages might be okay, too for average use.
Design +1

AkihiroSuda commented:
Design SGTM :+1:

thaJeztah commented:
> One thing to think about here is logger.Message could be any size, it just depends on how long the log line was...

Perhaps silly comment, but https://github.com/docker/docker/pull/22982/files#diff-0e0dba78caf4a57795625b5ce9943c9eR13 introduced a buffer size, i.e., messages longer than that size are split (even if there's no newline); or is that not relevant here?

cpuguy83 commented:
That's true, there is a max. But I think people want to control the buffer based on size not number of messages.
So I've pushed a new implementation doing just this.

cpuguy83 commented:
@justincormack Implementing this in `github.com/pkg/iotutils.BytesPipe` is fairly trivial I think, but wondering what's better, drop random byte slices or full messages.

cpuguy83 commented:
Although certainly implementing in BytesPipe is significantly faster since we don't need to do any extra allocations, where as each call to `Log()` as implemented here requires 1 allocation.

I went ahead and added an implementation in `BytesPipe` just to look at, presented without tests or benchmarks at the moment.

justincormack commented:
Best to drop full messages I think. They might be JSON or something and
truncating would be unhelpful.

On 26 Nov 2016 2:35 p.m., "Brian Goff" <notifications@github.com> wrote:

> @justincormack <https://github.com/justincormack> Implementing this in
> github.com/pkg/iotutils.BytesPipe is fairly trivial I think, but
> wondering what's better, drop random byte slices or full messages.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/28762#issuecomment-263066514>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdcPPpe7Uf5tLMOCojYaCpkvPg0em68ks5rCEOugaJpZM4K6tX6>
> .
>

AkihiroSuda commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,213 @@
+// the passed in logger.
+func NewRingLogger(driver Logger, logCtx Context, maxSize int64) Logger {
+	if maxSize < 0 {
+		maxSize = defaultRingMaxSize
```
nit: personally, I prefer zero-value for indicating the default

cpuguy83 replied:
Yes, but I wanted to support a 0 size, which would essentially mean store no more than 1 message.

AkihiroSuda commented on runconfig/opts/parse.go:
```diff
@@ -211,8 +211,8 @@ func AddFlags(flags *pflag.FlagSet) *ContainerOptions {
 	flags.StringVar(&copts.loggingDriver, "log-driver", "", "Logging driver for the container")
-	flags.StringVar(&copts.volumeDriver, "volume-driver", "", "Optional volume driver for the container")
 	flags.Var(&copts.loggingOpts, "log-opt", "Log driver options")
+	flags.StringVar(&copts.volumeDriver, "volume-driver", "", "Optional volume driver for the container")
```
unintended change, perhaps?

cpuguy83 replied:
Intended, just re-ordered so log-opt is underneath log-driver.

cpuguy83 commented:
Removed the commit with the bytespipe changes.

LK4D4 commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -176,7 +176,9 @@ func (l *logStream) Log(msg *logger.Message) error {
 		// buffer up the data, making sure to copy the Line data
-		l.messages <- logger.CopyMessage(msg)
+		var copy logger.Message
+		copy = *msg
```
This doesn't clone underlying message([]byte), which was whole point IIRC.

LK4D4 replied:
Copier uses parts of its own buffer for Line:
https://github.com/docker/docker/blob/master/daemon/logger/copier.go#L81
And drivers like fluentd can collect messages for later send, so their Line will be overriden by copier.
Also, I believe this was in comment to CopyMessage :D

cpuguy83 replied:
Oh you're right. I'll bring this back.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,213 @@
+package logger
+
+import (
+	"context"
```
should be net/context

LK4D4 commented:
Cond based buffer is kinda cool. Will review later today.

cpuguy83 commented:
Benchmarks:

```
BenchmarkRingLoggerThroughputNoReceiver-8           	300000000	       135 ns/op	 229.05 MB/s	     133 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputWithReceiverDelay0-8   	200000000	       212 ns/op	 146.06 MB/s	     113 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay1-8        	300000000	       130 ns/op	 237.55 MB/s	     133 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay10-8       	300000000	       132 ns/op	 234.68 MB/s	     133 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay50-8       	300000000	       126 ns/op	 244.84 MB/s	     133 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay100-8      	300000000	       131 ns/op	 235.54 MB/s	     133 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay300-8      	300000000	       130 ns/op	 237.30 MB/s	     133 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay500-8      	300000000	       128 ns/op	 240.39 MB/s	     133 B/op	       1 allocs/op
```

The happiest case (no receiver delay) is significantly slower. I think this is due to lock contention, but probably does not matter much unless the log driver is doing it's own queueing.
Even with a 1ms delay, throughput is back up.

cpuguy83 commented:
Updated benchmarks after fixing copy logic.
The no-delay scenario is really slow here.
Specifically, when I use a sync.Pool to handling messages when copying/done with messages, it drops from around 125MB/sec to 75MB/sec, but the other cases go from ~170MB/sec to 220MB/sec

```
BenchmarkRingLoggerThroughputNoReceiver-8           	300000000	       146 ns/op	 211.68 MB/s	      69 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputWithReceiverDelay0-8   	100000000	       388 ns/op	  79.84 MB/s	      52 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay1-8        	300000000	       170 ns/op	 181.94 MB/s	      69 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay10-8       	300000000	       153 ns/op	 201.78 MB/s	      69 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay50-8       	300000000	       138 ns/op	 224.37 MB/s	      69 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay100-8      	300000000	       146 ns/op	 211.85 MB/s	      69 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay300-8      	300000000	       143 ns/op	 216.44 MB/s	      69 B/op	       1 allocs/op
BenchmarkRingLoggerThroughputConsumeDelay500-8      	300000000	       144 ns/op	 214.35 MB/s	      69 B/op	       1 allocs/op
```

cpuguy83 commented:
I think I'm going make some changes to the copier.

cpuguy83 commented:
Ok, what I've done is gone ahead and make the copier handle the allocations.
This simplifies some things for log drivers that are buffering (and I'm not confident that prior to this PR we are doing the right thing w.r.t. making copies of the log message in buffered drivers).

It evens out the performance of the ring buffer but causes more allocations and more bytes being moved per op.
Still the performance here is a bit better than before #22982.
For comparison:

### Before #22982 (<= 1.12)
| Size | Ops | Speed | Memory | Allocs |
| --- | --- | --- | --- | --- |
| 1K | 10000 | 141169 ns/op | 16351 B/op | 39 allocs/op |
| 2K | 20000 | 98573 ns/op | 26577 B/op | 39 allocs/op |
| 4K | 10000 | 105285 ns/op | 47058 B/op | 39 allocs/op |
| 8K | 10000 | 138987 ns/op | 129302 B/op | 59 allocs/op |
| 16K | 10000 | 171856 ns/op | 294591 B/op | 100 allocs/op |
| 32K | 5000 | 307481 ns/op | 624206 B/op | 152 allocs/op |
| 64K | 2000 | 582192 ns/op | 1283438 B/op | 245 allocs/op |
| 128K | 1000 | 1212861 ns/op | 2601906 B/op | 422 allocs/op |
| 256K | 1000 | 2267578 ns/op | 5238848 B/op | 766 allocs/op |

### After #22982 (1.13.0-rcX)
| Size | Ops | Speed | Memory |Allocs |
| --- | --- | --- |  --- |  --- |
| 1K | 10000 | 101876 ns/op |  17664 B/op |  20 allocs/op | 
| 2K | 10000 | 100582 ns/op |  17664 B/op |  20 allocs/op | 
| 4K | 10000 | 103867 ns/op |  17664 B/op |  20 allocs/op | 
| 8K | 10000 | 117245 ns/op |  17664 B/op |  20 allocs/op | 
| 16K | 10000 | 120569 ns/op |  17664 B/op | 20 allocs/op | 
| 32K | 5000 | 236115 ns/op | 17664 B/op |  20 allocs/op | 
| 64K | 3000 | 434702 ns/op |  17664 B/op |  20 allocs/op |
| 128K | 2000 | 839920 ns/op |  17664 B/op |  20 allocs/op | 
| 256K | 1000 | 1611549 ns/op |  17664 B/op |  20 allocs/op |

### This PR
| Size | Ops | Speed | Memory |Allocs |
| --- | --- | --- | --- | --- |
| 1K                  |                     300000         |    40589 ns/op      |     28800 B/op   |      39 allocs/op |
| 2K               |                      300000       |      56270 ns/op     |      39040 B/op   |      39 allocs/op |
| 4K                  |                    100000         |   129520 ns/op       |    59520 B/op   |     39 allocs/op |
| 8K                |                       100000       |     154120 ns/op     |     100480 B/op   |      39 allocs/op |
| 16K           |                           100000       |     238414 ns/op    |      182400 B/op  |       39 allocs/op |
| 32K           |                            30000       |     448035 ns/op    |      347200 B/op  |       59 allocs/op |
| 64K            |                           20000       |     809671 ns/op   |      676800 B/op  |       99 allocs/op |
| 128K            |                          10000       |    1547572 ns/op     |    1336000 B/op |       179 allocs/op |
| 256K               |                        5000       |    3110902 ns/op    |     2654400 B/op        |339 allocs/op |

|       Delay          |       Ops    |     Speed       |      Throughput       |    Memory     |    Allocs         | 
|-----------------|-----------|------------|-------------|---------|-------------| 
| NoReceiver      | 200000000 | 87.9 ns/op | 352.59 MB/s | 37 B/op | 0 allocs/op | 
| ConsumeDelay0   | 100000000 | 152 ns/op  | 203.78 MB/s | 20 B/op | 0 allocs/op | 
| ConsumeDelay1   | 200000000 | 90.5 ns/op | 342.47 MB/s | 37 B/op | 0 allocs/op | 
| ConsumeDelay10  | 200000000 | 86.1 ns/op | 359.98 MB/s | 37 B/op | 0 allocs/op | 
| ConsumeDelay50  | 200000000 | 85.1 ns/op | 364.14 MB/s | 37 B/op | 0 allocs/op | 
| ConsumeDelay100 | 200000000 | 82.3 ns/op | 376.85 MB/s | 37 B/op | 0 allocs/op | 
| ConsumeDelay300 | 200000000 | 81.4 ns/op | 380.61 MB/s | 37 B/op | 0 allocs/op | 
| ConsumeDelay500 | 200000000 | 78.3 ns/op | 396.16 MB/s | 37 B/op | 0 allocs/op |

cpuguy83 commented:
And adding a pool in can bring allocs back down to 19 as before this PR and significantly faster *if* the consuming log driver put the message back into the pool.. but each log driver will have to put the message back.

cpuguy83 committed:
Use sync.Pool for logger Messages

This reduces allocs and bytes used per log entry significantly as well
as some improvement to time per log operation.

Each log driver, however, must put messages back in the pool once they
are finished with the message.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
This should be all good to go.
There are 2 commits, the first one adds the message ring buffer, the 2nd one makes the copier use a sync.Pool to deal with performance concerns.

thaJeztah commented:
ping @LK4D4 @AkihiroSuda PTAL

vdemeester commented:
@cpuguy83 needs a rebase 👼 
/cc @tonistiigi @LK4D4 ptal 🙇

cpuguy83 commented:
rebased

aaronlehmann commented on container/container.go:
```diff
@@ -38,6 +38,7 @@ import (
 	"github.com/docker/docker/runconfig"
 	"github.com/docker/docker/volume"
 	"github.com/docker/go-connections/nat"
+	units "github.com/docker/go-units"
```
I think `units` is unnecessary here.

aaronlehmann commented on daemon/logger/awslogs/cloudwatchlogs.go:
```diff
@@ -176,7 +176,7 @@ func (l *logStream) Log(msg *logger.Message) error {
 	if !l.closed {
 		// buffer up the data, making sure to copy the Line data
-		l.messages <- logger.CopyMessage(msg)
+		l.messages <- msg
```
The comment says the data is being copied, but it's no longer being copied.

Is it okay to send `msg` over the channel without copying it? A caller might call `Log` and then modify `msg` afterwards.

cpuguy83 replied:
Yes, I can remove this comment as it's no longer neccessary after the 2nd commit.

aaronlehmann commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,219 @@
+// This is used when lossy logging is OK.
+type RingLogger struct {
+	buffer *messageRing
+	ctx    context.Context
```
Since this is not taking advantage of context features such as hierarchy or value storage, could it just be a regular channel? I've heard contexts inside structures derided as an anti-pattern, though I admit to doing it sometimes.

cpuguy83 replied:
Yes I've heard this as well, but is really nice here because I don't have to worry about concurrent closes.
Probably have to add a separate mutex (or a `sync.Once`) just to handle close.

But :man_shrugging:
In the end, either work for me.

LK4D4 replied:
If we look closer on ctx usage here, we'll see that we don't even need channel. It can be variable with mutex. At least I don't see real selects which we can cancel with context or channel.

cpuguy83 replied:
That's true, it's nice on the for { select }, but since we're only selecting to check for closed I guess it makes sense.

aaronlehmann commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,219 @@
+				}
+			}
+		}
+	}()
```
nit: Maybe `go run()` instead of this closure? That would make it clearer that `newRingLogger` starts a goroutine, and simplify this function.

aaronlehmann commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,219 @@
+		// dump a message from the queue and try again
+		msg := r.queue[0]
+		r.queue = r.queue[1:]
+		r.sizeBytes -= int64(msg.Size)
```
Curious why this drops old messages instead of new ones.

cpuguy83 replied:
Good question. Is there a reason to do one vs the other?
Code wise, it's a bit simpler to drop the first element rather than the last.

aaronlehmann replied:
I'd think the simplest thing to do is just drop the message we're trying to enqueue if there's no space.

cpuguy83 replied:
Oh good point. 💡

cpuguy83 replied:
👍 This gives 2x throughput increase.

aaronlehmann commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	msg := r.queue[0]
+	r.queue = r.queue[1:]
```
This will be more alloc-heavy than a circular ringbuffer, but I'm not sure whether we care.

cpuguy83 replied:
I started looking at this, but I think the complexity will far outweigh any benefit.
Allocs don't happen enough to show up on benchmark results as is.

aaronlehmann commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,219 @@
+	ls := make([]*Message, 0, len(r.queue))
+	ls = append(ls, r.queue...)
+	r.sizeBytes = 0
+	r.queue = make([]*Message, 0, 100)
```
Why 100? `newRing` seems to use 1000.

cpuguy83 commented on container/container.go:
```diff
@@ -349,7 +350,23 @@ func (container *Container) StartLogger() (logger.Logger, error) {
 	}
-	return c(info)
+
+	l, err := c(info)
```
"creator" I believe,but yes.

cpuguy83 commented on daemon/logger/logger.go:
```diff
@@ -26,33 +26,44 @@ const (
+func (m *Message) Reset() {
+	m.Line = m.Line[:0]
+	m.Source = ""
+	for k := range m.Attrs {
```
Not really, because 1) it's not actually used except for reads (and only sometimes), and 2) it adds an extra allocation

cpuguy83 replied:
I'll just set this nil here.

cpuguy83 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,219 @@
+	wait *sync.Cond
+
+	sizeBytes int64 // current buffer size
+	cap       int64 // max buffer size size
```
But it's so pretty... I can change 😉

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+func (r *RingLogger) setClosed() {
+	atomic.StoreInt32(&r.close, 1)
```
maybe I'm missing something, but this looks like used only in tests.

cpuguy83 replied:
Yeah, forgot to add it in `Close()`

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,210 @@
+	buffer  *messageRing
+	l       Logger
+	logInfo Info
+	close   int32
```
hehe, another reserved name :)

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,210 @@
+)
+
+const (
+	defaultRingMaxSize = 1000000 // 1MB
```
It's kinda hard to count zeroes. I think 1024 * 1024 would be simpler to read. Also more scientific :microscope:

aaronlehmann replied:
Or 1e6

cpuguy83 commented:
Updated again.

LK4D4 commented on daemon/logger/fluentd/fluentd.go:
```diff
@@ -153,7 +153,9 @@ func (f *fluentd) Log(msg *logger.Message) error {
 	// fluent-logger-golang buffers logs from failures and disconnections,
 	// and these are transferred again automatically.
-	return f.writer.PostWithTime(f.tag, msg.Timestamp, data)
+	err := f.writer.PostWithTime(f.tag, msg.Timestamp, data)
```
I wonder if it makes sense to copy required fields and `PutMessage` before `PostWithTime`.
Probably not a big deal.

LK4D4 commented on daemon/logger/logger.go:
```diff
@@ -26,33 +26,42 @@ const (
+// Reset sets the message back to default values
+// This is used when putting a message back into the message pool.
+// Any changes to the `Message` struct should be reflected here.
+func (m *Message) Reset() {
```
Should be private I think.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+	}
+
+	if r.closeErr != nil {
+		r.mu.Unlock()
```
sounds like case for defer.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+
+	r.mu.Lock()
+	if r.closeErr != nil {
+		r.mu.Unlock()
```
definitely case for defer :)

cpuguy83 replied:
It's 1/4 the speed with defer.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+	ls := make([]*Message, 0, len(r.queue))
+	ls = append(ls, r.queue...)
+	r.sizeBytes = 0
+	r.queue = make([]*Message, 0, 1000)
```
Is buffer reusable? If we call Drain only after Close, then probably there is no point to recreate slice here. Maybe just `r.queue = r.queue[:0]`.

LK4D4 commented on daemon/logger/syslog/syslog.go:
```diff
@@ -132,10 +132,14 @@ func New(info logger.Info) (logger.Logger, error) {
+		err = s.writer.Info(string(msg.Line))
 	}
-	return s.writer.Info(string(msg.Line))
+	logger.PutMessage(msg)
```
here, copy of `msg.Line` and `PutMessage` early could reduce diff :)

LK4D4 commented:
@cpuguy83 some minor nits. Otherwise looks good.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+	sizeBytes int64 // current buffer size
+	maxBytes  int64 // max buffer size size
+	queue     []*Message
+	closeErr  error // error set on close
```
this variable is slightly confusing. It's actually boolean by its nature. I'd prefer to have boolean and return `errClosed` if it became true.

cpuguy83 commented:
This is updated again.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+}
+
+func newRing(maxBytes int64) *messageRing {
+	queueSize := 1000
```
btw if it would be message count limit and not size, this queue might be lock-free and alloc-free :)

cpuguy83 replied:
Yeah, but it's weird to do it based on number of messages since log messages will be all different sizes.
I would like to improve the case where there is effectively nothing being buffered, though... this is currently 1/4 the speed of cases where items are getting buffered due to lock contention.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+	for _, msg := range r.buffer.Drain() {
+		if err := r.l.Log(msg); err != nil {
+			logrus.WithField("driver", r.l.Name()).WithField("container", r.logInfo.ContainerID).Errorf("Error writing log message: %v", r.l)
+			break
```
we probably should put rest of the messages in pool anyway.

cpuguy83 replied:
You mean continue trying to log?

LK4D4 replied:
I don't know :/ maybe to iterate rest and PutMessage.

LK4D4 commented on daemon/logger/ring.go:
```diff
@@ -0,0 +1,211 @@
+// If the message is too big for the buffer it drops the oldest messages to make room
+// If there are no messages in the queue and the message is still too big, it adds the message anyway.
+func (r *messageRing) Enqueue(m *Message) error {
+	m.Size = len(m.Line) // ensure size is set
```
Why is this important? And why don't just use `mSize` here?

cpuguy83 replied:
I guess it's not really. Just always have a bit of twitch when calling a function vs just reading a param. 
Probably because method dispatch is so slow in ruby :)

LK4D4 replied:
I mean it's weird that you use both int64(m.Size) on 157 and mSize on 150.
Maybe it should be just `mSize := int64(len(m.Line))`

LK4D4 replied:
Field indeed looks weird instead of just len(m.Line) everywhere, but I'm ok with it if you like it.

cpuguy83 commented:
Updated and green.

LK4D4 reviewed (approved):
LGTM

aaronlehmann commented:
LGTM

LK4D4 commented:
ping @thaJeztah for docs review

thaJeztah reviewed (approved):
docs LGTM

@cpuguy83 will you open a pr against the vnext-engine branch in https://github.com/docker/docker.github.io for the additional docs changes?

Also ping @albers @sdurrheimer in case we have completion scripts for these options (`--log-opt`)

glasser commented:
Based on reading the source, it looks like there's a serious bug here in the journald and syslog drivers. (Haven't managed to verify yet with an actual built version.)  In both of those drivers, you call `logger.PutMessage(msg)` and then afterwards read `msg.Source`.  But PutMessage calls `msg.reset()` which clears `Source`. Am I missing something or do the drivers not need to save those values before calling PutMessage?

glasser commented:
I agree. But look at this code this PR put into [the syslog driver](https://github.com/moby/moby/blob/33fd3817b0f5ca4b87f0a75c2bd583b4425d392b/daemon/logger/syslog/syslog.go#L134-L141), which remains there to this day:

```
func (s *syslogger) Log(msg *logger.Message) error {
	line := string(msg.Line)
	logger.PutMessage(msg)
	if msg.Source == "stderr" {
		return s.writer.Err(line)
	}
	return s.writer.Info(line)
}
```

That sure seems to read from Source after calling PutMessage. Similar code exists in the journald driver.

cpuguy83 commented:
@glasser Yeah, that looks like an issue!
Do you want to open a PR to fix this?

glasser commented:
@cpuguy83 sure, #33832
